LoRa.Message/PHYPayload.cs

```diff
@@ -38,6 +40,134 @@ public PHYPayload(byte[] packet, byte[] nwkSKey, byte[] appSKey, byte[] appKey,
+                    System.Buffer.BlockCopy(payload.DevAddr.RawData.ToArray(), 0, cmac, Mhdr.RawData.Length + payload.AppNonce.RawData.Length + payload.NetID.RawData.Length, payload.DevAddr.RawData.Length);
+                    System.Buffer.BlockCopy(payload.DLSettings.RawData.ToArray(), 0, cmac, Mhdr.RawData.Length + payload.AppNonce.RawData.Length + payload.NetID.RawData.Length + payload.DevAddr.RawData.Length, payload.DLSettings.RawData.Length);
+                    System.Buffer.BlockCopy(payload.RxDelay.RawData.ToArray(), 0, cmac, Mhdr.RawData.Length + payload.AppNonce.RawData.Length + payload.NetID.RawData.Length + payload.DevAddr.RawData.Length + payload.DLSettings.RawData.Length, payload.RxDelay.RawData.Length);
+                    return cmac;
+                    break;
+                case MessageType.Data:
+                    throw new NotImplementedException();
+                    break;
+                case MessageType.Proprietary:
+                    throw new NotImplementedException();
+                default:
+                    throw new NotImplementedException();
+            }
+        }
+
+          byte[] SKey()
+        {
+            switch (Mhdr.MessageType)
+            {
+                case MessageType.JoinRequest:
+                    throw new NotImplementedException();
+                    break;
+                case MessageType.JoinAccept:
+                    JoinAcceptMessage payload = (JoinAcceptMessage)MacPayload;
+                    byte[] skey = new byte[payload.AppNonce.RawData.Length +
+                    payload.NetID.RawData.Length +
+                   (payload.joinRequestMessage.DevNonce.Value.Length / 2) + ("00000000000000".Length / 2)];
+
+                    System.Buffer.BlockCopy(payload.AppNonce.RawData.ToArray(), 0, skey, 0, payload.AppNonce.RawData.Length);
+                    System.Buffer.BlockCopy(payload.NetID.RawData.ToArray(), 0, skey, payload.AppNonce.RawData.Length, payload.NetID.RawData.Length);
+                    System.Buffer.BlockCopy(payload.joinRequestMessage.DevNonce.Value.FromHexString().Reverse().ToArray(), 0, skey, payload.AppNonce.RawData.Length + payload.NetID.RawData.Length, payload.joinRequestMessage.DevNonce.Value.Length / 2);
+
+                    System.Buffer.BlockCopy(("00000000000000").FromHexString(), 0, skey, payload.AppNonce.RawData.Length + payload.NetID.RawData.Length + (payload.joinRequestMessage.DevNonce.Value.Length / 2), "00000000000000".Length / 2);
+                    return skey;
+                    break;
+                case MessageType.Data:
+                    throw new NotImplementedException();
+                    break;
+                case MessageType.Proprietary:
+                    throw new NotImplementedException();
+                default:
+                    throw new NotImplementedException();
+            }
+        }
+
+        public byte[] ToBytes( byte[] AppKey, byte[] appKey)
```

Comment:
If I understand correctly, this method create a new payload? I think this will be better use extension methods for this.